Disable Metrics/BlockLength on specific repositories

Polo2 opened the pull request:
**Proposition:** I think that cop `Metrics/BlockLength` should be disabled for these repositories:
- `spec/` : not only spec/**/*.rb ,
- `serializers/` : one line per CSV make this rule to constraignant
- `tasks/` : Some tasks can be nested in one block, * 25 lines is not enough in this case.

Polo2 committed:
Disable Metrics/ from any repositories

Cop 'Metrics/BlockLength' should be disabled for repositories:
default.yml:
- spec/ (not only spec/**/*.rb)
rails.yml:
- app/serializers/ : one line per CSV make this rule to constraignant
- lib/tasks/ : Some tasks can be nested in one block. 25 lines is not enough

tjouan commented on config/default.yml:
```diff
@@ -30,6 +30,10 @@ Metrics:
 Metrics/BlockLength:
   Max: 25
+  Exclude:
+    - spec/
```
I agree there is a problem, but before we merge this, I'd like to understand
why the config for `Metrics` does not apply to `Metrics/BlockLength`, isn't it
supposed to behave this way? Should we remove it on `Metrics` first?

https://github.com/alpinelab/codestyle/blob/588a78d14775214cc3619463c0499047a03269b0/config/default.yml#L27-L29

---

Does rubocop handles `spec/` differently than `spec`? If the answer is no, I
think we should favor just `spec`.

tjouan replied:
I think @Polo2 had the issue in files ending with `spec/**/*.rb` files, because
I noticed and reported the same issue some time ago (not on GitHub however).

tjouan commented on config/default.yml:
```diff
@@ -30,6 +30,10 @@ Metrics:
   Max: 25
+  Exclude:
+    - spec/
+    - app/serializers/
```
This should be in `config/rails.yml` config file.

Polo2 replied:
I agree, but `rails.yml` seems to override `default.yml` on this exlusion list

tjouan commented on config/default.yml:
```diff
@@ -30,6 +30,10 @@ Metrics:
+  Exclude:
+    - spec/
+    - app/serializers/
+    - lib/tasks/
```
This should be in `config/rails.yml` config file.

Polo2 replied:
I agree, but `rails.yml` seems to override `default.yml` on this exlusion list

tjouan commented:
Globally, I'm favorable to those changes, but in a different form (especially
moving Rails specifics to its config).

michaelbaudino commented on config/default.yml:
```diff
@@ -30,6 +30,10 @@ Metrics:
 Metrics/BlockLength:
   Max: 25
+  Exclude:
+    - spec/
```
Agreed: this is strange and should be investigated :male_detective: Maybe you had offenses on files with another extension than `.rb`, @Polo2 ?

Polo2 replied:
I did investigated :male_detective: !
Issue don't comes from `spec/**/*.rb` to `spec/`, issue comes from how keys are merged... or **not merged !**

On `rails.yml`, I realized that:
```ruby
Metrics/BlockLength:
  Exclude:
    - config/routes.rb
```
overrides informations wich comes from on `config.yml/`:
```ruby
Metrics:
  Exclude:
    - spec/**/*.rb
```

I don't know how to fix this .... Could it come from: 
```ruby
inherit_mode:
  merge:
    - Exclude
```
?

michaelbaudino replied:
Well,
```
inherit_mode:
  merge:
    - Exclude
```
makes Rubocop merges values passed to `Exclude`.

Here, it might not be handled by inheritance but by because `Metrics` is the namespace that contains `Metrics/BlockLength`.

I can't find anything in the documentation about that, but what about trying a cop-specific inheritance mode like:

```yaml
Metrics/BlockLength:
  inherit_mode:
    merge:
      - Exclude
  Exclude:
    - config/routes.rb
```

If it does not work, you may have to use the not-very-DRY version:

```yaml
Metrics/BlockLength:
  Exclude:
    - spec/**/*.rb
    - config/routes.rb
```

michaelbaudino commented:
I'll vote to include those changes too, once they've been reworked :+1:

michaelbaudino commented on config/default.yml:
```diff
@@ -30,6 +30,10 @@ Metrics:
+  Exclude:
+    - spec/
+    - app/serializers/
+    - lib/tasks/
```
It does not override it: it merges it. Thanks to this configuration: https://github.com/alpinelab/codestyle/blob/bd9fd8d4ba2351b1d45941f586a4b9a9d73915f4/config/default.yml#L3-L5

michaelbaudino commented:
The commit message mentions changes in `default.yml` but the commit does not actually contain any change on this file :thinking:

Polo2 committed:
Avoid deep merging issue

That's an issue on Rubobop: Exclude lists aren't deep merged.

Indeed, Exclude list for Metrics on default.yml was overriding
Exclude list for Metrics/BlockLength on Rails.yml.

Polo2 commented:
Related issues on Rubocop repository:
- https://github.com/rubocop-hq/rubocop/issues/4477
- https://github.com/rubocop-hq/rubocop/issues/4621
- https://github.com/rubocop-hq/rubocop/issues/5480

michaelbaudino commented:
I'm actually starting to wonder if setting `Exclude` for a whole namespace is something that Rubocop expects :thinking: I see not example in Rubocop documentation nor in the project issues that does something similar to https://github.com/alpinelab/codestyle/blob/01a96db2631d3808880efffeb071b054d7ed5f92/config/default.yml#L27-L29

(except for special cops `Enabled` and `EnabledByDefault`)

michaelbaudino commented:
It might actually be worth creating an issue on Rubocop project to highlight our problem and the fact that there seems to be inheritance and different inheritance modes for files inheritance, but not for namespace inheritance.
